File: src/windows/paths.c
#include "paths.h"
#include "../crt.h"
#include "../util/logging.h"

DoorstopPaths *paths_init(void *doorstop_module, bool_t fixed_cwd) {
    char_t *app_path = program_path();
    char_t *app_dir = get_folder_name(app_path);
    char_t *working_dir = get_working_dir();
    char_t *doorstop_path = NULL;
    get_module_path(doorstop_module, &doorstop_path, NULL, 0);

    char_t *doorstop_filename = get_file_name(doorstop_path, FALSE);

    LOG("Doorstop started!");
    LOG("Executable path: %" Ts, app_path);
    LOG("Application dir: %" Ts, app_dir);
    LOG("Working dir: %" Ts, working_dir);
    LOG("Doorstop library path: %" Ts, doorstop_path);
    LOG("Doorstop library name: %" Ts, doorstop_filename);

    if (fixed_cwd) {
        log_warn("Working directory was not the same as app directory, fixed "
                 "it automatically.");
    }

    DoorstopPaths *paths = malloc(sizeof(DoorstopPaths));
    paths->app_path = app_path;
    paths->app_dir = app_dir;
    paths->working_dir = working_dir;
    paths->doorstop_path = doorstop_path;
    paths->doorstop_filename = doorstop_filename;
    return paths;
}

void paths_free(DoorstopPaths *const paths) {
    free(paths->app_path);
    free(paths->app_dir);
    free(paths->working_dir);
    free(paths->doorstop_path);
    free(paths->doorstop_filename);
}

File: src/windows/paths.zig
const std = @import("std");

const root = @import("../root.zig");
const alloc = root.alloc;
const logger = root.logger;
const util = root.util;

const os_char = util.os_char;
const c_bool = util.c_bool;

app_path: [:0]os_char,
app_dir: [:0]os_char,
working_dir: [:0]os_char,
doorstop_path: [:0]os_char,

pub fn init() @This() {
    const app_path = util.paths.programPath();
    const app_dir = util.paths.getFolderName(app_path);
    const working_dir = util.paths.getWorkingDir();
    const doorstop_path_buf = util.paths.getModulePath(root.entrypoint.windows.doorstop_module).?;
    defer doorstop_path_buf.deinit();
    const doorstop_path = alloc.dupeZ(os_char, doorstop_path_buf.result) catch @panic("Out of memory");

    logger.debug("Doorstop started!", .{});
    logger.debug("Executable path: {}", .{std.unicode.fmtUtf16Le(app_path)});
    logger.debug("Application dir: {}", .{std.unicode.fmtUtf16Le(app_dir)});
    logger.debug("Working dir: {}", .{std.unicode.fmtUtf16Le(working_dir)});
    logger.debug("Doorstop library path: {}", .{std.unicode.fmtUtf16Le(doorstop_path)});

    return .{
        .app_path = app_path,
        .app_dir = app_dir,
        .working_dir = working_dir,
        .doorstop_path = doorstop_path,
    };
}

pub fn deinit(self: *@This()) void {
    alloc.free(std.mem.span(self.app_path));
    alloc.free(std.mem.span(self.app_dir));
    alloc.free(std.mem.span(self.working_dir));
    alloc.free(std.mem.span(self.doorstop_path));
}
